docs: ✏️ Securing a node with SSL/TLS (#9)
fleek-network/fleek-network-docs: guides/Network nodes/fleek-network-securing-a-node-with-ssl-tls.md

category: Tutorial
tags:
- DCDN
- Guide
- Getting Started
- Fleek Network
- security
- SSL/TSL
- Let's Encrypt
- Certbot
- Docker
---

<!--
  The following import is intentional (see partial <CheckoutCommitWarning />)
-->
import CheckoutCommitWarning from '../partials/_checkout-commit-warning.mdx';
import Author from '@site/src/components/Author';

![](./assets/fleek-network-how-to-secure-a-network-node.png?202301251236)

## Introduction

Secure Sockets Layer or Transport Layer Security are the technologies used to keep an internet connection secure by protecting sensitive data that is transferred between systems. It's a protocol for servers and web clients to ensure that data passed between them is private!

A secure connection is required to prevent others from modifying, reading or tempering the data that is transmitted. Otherwise, it can be intercepted by third parties, such as hackers.
HTTP connections are insecure by default, SSL or TLS (the modern version of SSL) is used to encrypt the data by using advanced encryption algorithms in transit, preventing others from reading it as it's transferred over the connection.

> 💡 Generally referred to as SSL, or TLS (Transport Layer Security), the modern version that's usually named SSL)

Author: Done

In recent years, the non-profit organization [Let's Encrypt](https://en.wikipedia.org/wiki/Let's_Encrypt) started providing TLS certificate encryption, that is free of charge. The Let's Encrypt organization also improved the administration experience by providing automation features for system administrator processes.

In that respect, we'll look into how to secure a Fleek Node with TLS certificates provided by the nonprofit Certificate Authority (CA) Let's Encrypt, by harnessing the knowledge provided in our [Running a node in a Docker container](fleek-network-running-a-node-in-a-docker-container) guide. We'll look into how to set up our custom domain in the Nginx configuration, generate dummy certificates, and real certificates and understand how the certificate automation is handled to mitigate issues, such as expiration.

> The knowledge demonstrated here should be useful and can be easily applied to native setups (none Docker). As some troubleshooting and differences are found in different systems, we'll stick with our Docker Stack for that matter and simplicity!

Reviewer: ```suggestion
The knowledge demonstrated here should be useful and can be easily applied to native setups (not Docker). As some troubleshooting and differences are found in different systems, we'll stick with our Docker Stack for that matter and simplicity!
```

the `none` doesn't sound right here 🤔

As an example, let's say that we have a domain `fleek.xyz`, where we set up a subdomain `fleek-network-node` and that our server IP Address (where the Fleek Network Node is running from) is `181.196.118.156`.

The custom domain address would be:

```sh
fleek-network-node.fleek.xyz
```

💡 Keep a note about the domain name as it'll be required for your reverse proxy configuration (NGINX).

Our server IP Address (this should be public, accessible outside the private network, otherwise external Nodes won't be able to communicate):

```sh
181.196.118.156
```

Depending on the "domain name registry" dashboard, you'd have to update the `A Records`, here's an example:

| Type        | Host                              | Answer            | TTL         |
| ----------- | --------------------------------- | ----------------- | ----------- |
| A           | fleek-network-node.fleek.xyz      |  181.196.118.156  | 300         |

Learn more about record types [here](https://en.wikipedia.org/wiki/List_of_DNS_record_types).

DNS changes may take some time to propagate, thus make sure to [verify the DNS records](#how-to-verify-dns-records).

> ### Alternatively, getting a free domain name for testing?

Reviewer 2: ```suggestion
### Alternatively, how can one get a free domain name for testing?
```

Domain names are costly, you might not have one, or not be interested at this time! For this reason, we're going to provide some hints on how to set up a free domain name for your server.

> There are plenty of free domain server providers but for this guide will opt for [Ngrok](https://ngrok.com), there are other alternatives such as [Localhost.run](localhost.run) or [LocalTunnel.me](https://localtunnel.me), [Ngrok](https://ngrok.com/), but we're keeping things simple here and will not try to provide support for anything else, but should be quite similar! 

Reviewer 3: ```suggestion
There are plenty of free domain server providers, but for this guide we will opt for [Ngrok](https://ngrok.com). There are other alternatives such as [Localhost.run](localhost.run) or [LocalTunnel.me](https://localtunnel.me), but we're keeping things simple here and will not try to provide support for anything else. Using an alternative should be quite similar! 
```

### How to verify DNS records?

Once you have [set up the DNS records](#how-to-set-up-the-dns-settings-for-a-node-server), verify the records are set correctly, for that you can easily run `dig` CLI or website like [mxtoolbox](https://mxtoolbox.com/). As an example, we're going to use the `dig` CLI!

`Dig` (Domain Information Groper) is a network administration command-line tool for querying the Domain Name System (DNS). Read more about it [here](https://en.wikipedia.org/wiki/Dig_(command)).

```sh
dig fleek-network-node.fleek.xyz +nostats +nocomments +nocmd
```

```sh
;fleek-network-node.fleek.xyz.		IN	A
fleek-network-node.fleek.xyz.	390	IN	A	dyna.wikimedia.org.
```

👆 The goal is to determine if the DNS record matches the machine IP Address we set up earlier. There might be cases you'll need to wait longer, but it's good to double-check your DNS record dashboard and revise it if the desired settings fail to show!

> 

Reviewer 3: hmm doesn't seem to have the ip address in the example response

If you have followed our introductory guides, you have probably commented out or removed the SSL configuration, keep reading to find the template or check the repository `app.conf` as mentioned earlier.

⚠️ Beware, that this is highly customizable and we might make tweaks in any earlier version of `ursa` which that guide might not consider. Check our original repository for any updates! Also, some advanced users have a preference to modify this on their own. In any case, for more on this you have to check the [Nginx docs](https://docs.nginx.com/) to learn more!

2) Replace the pathname with your custom domain, by replacing the following `ursa.earth` domain with yours

```sh
# SSL code
ssl_certificate /etc/letsencrypt/live/ursa.earth/fullchain.pem;
ssl_certificate_key /etc/letsencrypt/live/ursa.earth/privkey.pem;
```

For our demo, we replace `ursa.earth` to `fleek-network-node.fleek.xyz`

```sh
# SSL code
ssl_certificate /etc/letsencrypt/live/fleek-network-node.fleek.xyz/fullchain.pem;
ssl_certificate_key /etc/letsencrypt/live/fleek-network-node.fleek.xyz/privkey.pem;
```

> 

Reviewer 3: ```suggestion

> Note: if multiple domain names are used, the `init-letsencrypt.sh` script will export them under the first domain provided

```

    ssl_certificate_key /etc/letsencrypt/live/<YOUR DOMAIN NAME>/privkey.pem;

    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

    location /stub_status {
      stub_status;
    }

    location / {
      add_header content-type  application/vnd.ipld.raw;
      add_header content-type  application/vnd.ipld.car;
      add_header content-type  application/octet-stream;
      add_header cache-control public,max-age=31536000,immutable;

      proxy_cache nodecache;
      proxy_cache_valid 200 31536000s;
      add_header X-Proxy-Cache $upstream_cache_status;
      proxy_cache_methods GET HEAD POST;
      proxy_cache_key "$request_uri|$request_body";
      client_max_body_size 1G;

      proxy_pass http://ursa:4069;
    }
}
```

### Running the Let's Encrypt automation script

First, change the directory to the `docker/full-node` path in your local repository. This is where our `init-letsencrypt.sh` is located!

```sh
cd docker/full-node
```

The directory file structure should look like:

```sh
.
├── README.md
├── data
├── docker-compose.yml
└── init-letsencrypt.sh

> ```

Reviewer 3: Maybe this bit of info could be better put farther above where the chmod +x is done

We have to provide a domain name, such as the example `fleek-network-node.fleek.xyz` and a valid email address `ops@fleek.xyz` that you as an operator have access to.

```sh
EMAIL="<VALID EMAIL>" DOMAINS="<DOMAIN WITH CORRECT DNS RECORDS>" ./init-letsencrypt.sh
```

Since we have already:

- Updated the `A Records` in the [previous step](#how-to-set-up-the-dns-settings-for-a-node-server)
- [Verified](#how-to-verify-dns-records) that the DNS records are correct

We can execute the `init-letsencrypt.sh` by prefixing the script with the variables `EMAIL` and `DOMAINS`.

- DOMAINS - A list separated by space, e.g. `DOMAINS="foobar.com my-node.fleek.xyz"`
- EMAIL - A valid email address that the server administrator, the ops operator or anyone of concern should have access to, e.g. `ops@fleek.xyz`

> 

Reviewer 3: for redundancy:
```suggestion

> Note: if multiple domains are provided, the certificates will be stored under the first one

```

These files will be updated when the certificate renews.

NEXT STEPS:
- The certificate will need to be renewed before it expires. Certbot can automatically renew the certificate in the background, but you may need to take steps to enable that functionality. See https://certbot.org/renewal-setup for instructions.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
If you like Certbot, please consider supporting our work by:
 * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
 * Donating to EFF:                    https://eff.org/donate-le
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

### Reloading nginx ...
2023/01/25 18:56:34 [notice] 17#17: signal process started
```

It's a long output but we can find the last few paragraphs useful, where it reads:

```sh
Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/fleek-network-node.fleek.xyz/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/fleek-network-node.fleek.xyz/privkey.pem
This certificate expires on 2023-04-25.
These files will be updated when the certificate renews.
```

> You may also find that the paragraph after, that mentions that the certificate expires is important. If you are running the Docker Stack we provide, the `certbot` is installed and automating the certification for you.

Reviewer 3: ```suggestion
You may also find that the paragraph after, that mentions that the certificate expires is important. If you are running the Docker Stack we provide, the `certbot` is installed and automates renewing the certification for you.
```
